aquadoggo/src/domain.rs

```diff
@@ -0,0 +1,1060 @@
+        backlink.map(|entry| entry.entry_bytes()).as_deref(),
+    )?;
+
+    ///////////////////////////////////
+    // ENSURE SINGLE NODE PER AUTHOR //
+    ///////////////////////////////////
+
+    // @TODO: Missing a step here where we check if the author has published to this node before, and also
+    // if we know of any other nodes they have published to. Not sure how to do this yet.
+
+    ///////////////////////////////
+    // VALIDATE OPERATION VALUES //
+    ///////////////////////////////
+
+    // @TODO: We skip this for now and will implement it in a follow-up PR
+    // validate_operation_against_schema(store, operation.operation()).await?;
+
+    //////////////////////////
+    // DETERINE DOCUMENT ID //
+    //////////////////////////
+
+    let document_id = match operation.action() {
+        OperationAction::Create => {
+            // Derive the document id for this new document.
+            entry_encoded.hash().into()
+        }
+        _ => {
+            // We can unwrap previous operations here as we know all UPDATE and DELETE operations contain them.
+            let previous_operations = operation.previous_operations().unwrap();
+
+            // Get the document_id for the document_view_id contained in previous operations.
+            // This performs several validation steps (check method doc string).
+            let document_id =
+                get_validate_document_id_for_view_id(store, &previous_operations).await?;
+
+            // Ensure the document isn't deleted.
+            ensure_document_not_deleted(store, &document_id)
+                .await
+                .map_err(|_| {
+                    "You are trying to update or delete a document which has been deleted"
+                })?;
+
+            document_id
+        }
+    };
+
+    // Verify the claimed log id against the expected one for this document id and author.
+    verify_log_id(store, &author, log_id, &document_id).await?;
+
+    ///////////////
+    // STORE LOG //
+    ///////////////
+
+    // If this is a CREATE operation it goes into a new log which we insert here.
+    if operation.is_create() {
+        let log = S::StorageLog::new(&author, &operation.schema(), &document_id, log_id);
+
+        store.insert_log(log).await?;
+    }
+
+    ///////////////////////////////
+    // STORE ENTRY AND OPERATION //
+    ///////////////////////////////
+
+    // Insert the entry into the store.
+    store
+        .insert_entry(S::StorageEntry::new(entry_encoded, operation_encoded)?)
+        .await?;
+    // Insert the operation into the store.
+    store
+        .insert_operation(
+            &S::StorageOperation::new(&author, &entry_encoded.hash().into(), &operation).unwrap(),
+            &document_id,
+        )
+        .await?;
+
+    /////////////////////////////////////
+    // DETERMINE NEXT ENTRY ARG VALUES //
+    /////////////////////////////////////
+
+    let next_seq_num = increment_seq_num(&mut seq_num.clone())?;
```

Comment:
This is an interesting scenario I've stumbled upon: You successfully published your last possible entry in the log (lets say you reached the u64::MAX) and then _still_ the request will fail as we reached the end. Is that correct? On one hand no, as the "request to publish" succeeded (the entry and operation made it into the database), on the other hand yes, the response can not be fullfilled.

Not sure what to do here ..! Send a "success" response but empty arguments?!

Reply:
Yeh, it's a funny one... Early I had the method calculate the next args _before_ storing the new entry and operation, so that if it errors, then it actually is an error and the insertion didn't happen. This could be a good patter, it would cut of the log one entry before max, but maybe that's a good thing?

Reply:
We can unwrap here safely. Let people have their max seq num!

Reply:
Do we have any consensus on this? I suggest:

- calculate `next_args` _before_ storing the new entry and if we find out we're at the MAX_SEQ_NUM then we already error (one seq num early) and _don't_ store this entry. At least for now I think this is a fine pattern. Obvs some way of handling a "warning" but successful publishing of entry would be better later on.

Reply:
ping @adz and @cafca about the above

Reply:
> Do we have any consensus on this? I suggest:
> 
>     * calculate `next_args` _before_ storing the new entry and if we find out we're at the MAX_SEQ_NUM then we already error (one seq num early) and _don't_ store this entry. At least for now I think this is a fine pattern. Obvs some way of handling a "warning" but successful publishing of entry would be better later on.

That would mean you can't store your last entry even though you technically could, right? I think thats an okay temporary solution for now as probably reaching the last log id / seq num is highly unlikely (therefore a rare edge case), technically I think we could find something better, like thinking about an alternative "next entry args" response for publish entry in these cases (which is not an request error, but different JSON).